File: with_node/src/day3.ts
export function part1(content: string[]) {
  const matches = content.join("").matchAll(/mul\((\d{1,3}),(\d{1,3})\)/g);

  return [...matches].reduce(
    (sum, [_, a, b]) => sum + Number(a) * Number(b),
    0
  );
}

export function part2(content: string[]) {
  const matches = content
    .join("")
    .matchAll(/(don't|do)\(\)|mul\((\d{1,3}),(\d{1,3})\)/g);

  return [...matches].reduce(
    (acc, [_, instruction, a, b]) => {
      if (instruction) {
        return { sum: acc.sum, processing: instruction === "do" };
      }
      if(acc.processing && a && b) {
        return { sum: acc.sum + Number(a) * Number(b), processing: true }
      }
      return acc
    },
    { processing: true, sum: 0 }
  ).sum;
}

File: 2024/with_python/src/day3.py
import re


def part1(content: list[str]) -> int:
    matches = re.finditer(r"mul\((\d{1,3}),(\d{1,3})\)", "".join(content))
    return sum(int(a) * int(b) for match in matches for a, b in [match.groups()])


def part2(content: list[str]) -> int:
    matches = re.finditer(
        r"(don't|do)\(\)|mul\((\d{1,3}),(\d{1,3})\)", "".join(content)
    )
    processing = True
    sum = 0

    for match in matches:
        instruction, a, b = match.group(1), match.group(2), match.group(3)

        if instruction in ("do", "don't"):
            processing = instruction == "do"

        if processing and a and b:
            sum += int(a) * int(b)

    return sum
